### PR TITLE
feat(qwen-image-edit-plus): condition prompt embeds on the source image (trainside)

### PR DESCRIPTION
**Pull request by CjhHa1**
## Maintainer Update (2026-07-26)

Revalidated against current `main`: merge-tree clean, full pre-commit clean, recipe composition clean, and an out-of-tree inline parity harness passed for upstream dimension calculation, masked-hidden extraction, multimodal processor/text-encoder plumbing, and positive/negative CFG image routing. No test file was added. The author’s subsequent 8×H20 one-rollout trainside smoke is reflected below.

## Summary

Follow-up to #128. Trainside Qwen-Image-Edit-Plus encoded prompts **text-only**, omitting the multimodal prompt conditioning that upstream `QwenImageEditPlusPipeline` (and the SGLang rollout path) use. The source image only reached the VAE latent-concat path, never the Qwen2.5-VL text encoder, so trainside rollout/replay conditioned on a weaker distribution than the Edit-Plus checkpoint was trained/served with. (This is exactly the "source image not in text encoder" gap discussed on #128.)

This PR adds `QwenImageEditPlusTextEmbedStage`, which mirrors upstream `_get_qwen_prompt_embeds`:
- prefixes the edit chat template with `Picture 1: <|vision_start|><|image_pad|><|vision_end|>`;
- runs the Qwen2.5-VL processor (`pixel_values` + `image_grid_thw`) on the source image resized to the ~384² *condition* grid (distinct from the ~1024² *VAE* grid used for latent concat);
- drops the 64-token edit-template system prefix (vs 34 for base text-only).

Both the positive and negative (CFG) branches feed the same source image. Gated by `QwenImageEditPlusPipelineConfig.use_condition_image_prompt` (default `True`; `False` falls back to the base text-only stage).

Follow-up hardening after review:
- `from_config()` now passes `max_sequence_length`, `use_condition_image_prompt`, and the processor root derived from the text-encoder override; the loader uses the checkpoint’s fixed `processor/` subfolder.
- Shared `extract_masked_hidden` helper extracted from the base Qwen-Image text-embed stage.
- New `ImageConditionedEmbedStage` protocol for the multimodal embed signature.
- Trainside recipe `qwen_image_edit_plus_nft.yaml` explicitly sets / documents `use_condition_image_prompt: true`.
- Processor loading uses `Qwen2VLProcessor.from_pretrained(..., subfolder=...)` so Edit checkpoints without a root `model_type` still load.

## Related Issue

Follow-up to #128 (Qwen-Image-Edit-Plus). N/A otherwise.

## Test Plan

- `ruff check` / `ruff format --check` on the changed Python files: clean.
- `python -m compileall` on the changed files: clean.
- pre-commit (incl. `recipe _target_ paths resolve`): passed.
- Recipe parse check: `qwen_image_edit_plus_nft.yaml` resolves `use_condition_image_prompt: true` for both bundle and pipeline.
- **GPU embed parity vs upstream diffusers** (run locally on H20 with `/apdcephfs_fsgm3/.../Qwen/Qwen-Image-Edit`, bf16):
  - oracle: `diffusers.pipelines.qwenimage.pipeline_qwenimage_edit_plus.QwenImageEditPlusPipeline._get_qwen_prompt_embeds`
  - candidate: `QwenImageEditPlusTextEmbedStage.embed(texts, images)`
  - condition-image resize vs `oracle.image_processor.resize`: size `(512, 288)`, max abs pixel diff `0`
  - positive prompt: shape `(1, 205, 3584)`, mask tokens `205`, max abs embed diff `0`, cosine `1.0`
  - CFG empty negative `" "`: shape `(1, 193, 3584)`, mask tokens `193`, max abs embed diff `0`, cosine `1.0`
  - attention masks byte-identical in both cases
- End-to-end DiffusionNFT trainside smoke passed on 8×H20 with one rollout and one FSDP train step (`reward=0.7517`, `loss=8.6618`, `gn=0.0864`); details are in the PR comment.

## Compatibility / Risk

- **Default behavior change**: `use_condition_image_prompt` defaults to `True`, so the existing trainside recipe `qwen_image_edit_plus_nft.yaml` now feeds the image into the text encoder. The previously reported trainside reward curve was produced with the text-only path — set `use_condition_image_prompt: false` in the bundle config to reproduce it.
- **Assumptions confirmed on the local Edit checkpoint**: processor lives under the `processor/` subfolder; edit template + `drop_idx=64` match installed diffusers `QwenImageEditPlusPipeline`. The exact trainside recipe subsequently completed the one-rollout 8×H20 smoke described above.
- The SGLang path (`load_text_encoder=false`) is unaffected — the stage/processor are only built trainside when a text encoder is present. No checkpoint/data-format/API changes.

## Reviewer Notes

- AI-assisted implementation; grounded in upstream diffusers `pipeline_qwenimage_edit_plus.py::_get_qwen_prompt_embeds`.
- GPU embed parity vs upstream and the 8×H20 trainside one-rollout smoke are green.

## Checklist

- [x] I reviewed the changed code and removed unrelated/generated artifacts.
- [x] I updated tests, docs, and configs where needed, or explained why not.

### CODE REVIEW

**Comment by CjhHa1** [comment]
## Review follow-up + GPU embed parity

Pushed a follow-up commit addressing review items **#3 / #5 / #7 / #8**, plus a processor-loading fix found while running real-checkpoint parity:

- `from_config()` now forwards `max_sequence_length`, `use_condition_image_prompt`, `processor_path`, and `processor_subfolder`.
- Trainside recipe `qwen_image_edit_plus_nft.yaml` explicitly sets / documents `use_condition_image_prompt: true`.
- Shared `extract_masked_hidden` helper extracted from base Qwen-Image text-embed.
- Added `ImageConditionedEmbedStage` for the multimodal `embed(texts, images)` signature.
- Switched processor load to `Qwen2VLProcessor.from_pretrained(..., subfolder=...)` — `AutoProcessor` failed on the Edit checkpoint root because it has no `model_type`.

### Parity result (local H20, bf16, `Qwen-Image-Edit`)

Oracle = upstream `QwenImageEditPlusPipeline._get_qwen_prompt_embeds`; candidate = `QwenImageEditPlusTextEmbedStage`.

| Check | Result |
| --- | --- |
| Condition resize vs `image_processor.resize` | size `(512, 288)`, max abs pixel diff `0` |
| Positive embeds | shape `(1, 205, 3584)`, mask `205`, max abs `0`, cosine `1.0` |
| CFG negative `" "` | shape `(1, 193, 3584)`, mask `193`, max abs `0`, cosine `1.0` |
| Attention masks | byte-identical |

Still pending: end-to-end DiffusionNFT trainside smoke on `qwen_image_edit_plus_nft.yaml`.

**Comment by CjhHa1** [comment]
## Trainside e2e smoke (`qwen_image_edit_plus_nft.yaml`) — passed

Ran on Taiji worker `28.49.33.52` (8×H20), synthetic image-edit smoke set (`datasets/image_edit/smoke`, 8 samples), overrides: `num_rollouts=1 batch_size=8 samples_per_prompt=2 save_interval=0`.

**Result**
```
rollout 1/1  reward=0.7517  loss=8.6618 gn=0.0864 lr=3.00e-04
```

Path exercised: `QwenImageEditPlusBundle` → image-conditioned `QwenImageEditPlusTextEmbedStage` (`use_condition_image_prompt=true`) → `DiffusionNFT` trainside rollout + one train step (FSDP + PickScore).

Note: cold load from CephFS is slow (~12h wall for serialized 8-rank full TE/transformer materialize); once loaded, the NFT step itself finished cleanly.

**Review by leviking98z-rgb — APPROVED** [review]
